pkg/pgmodel/querier/row.go

```diff
@@ -0,0 +1,183 @@
+package querier
+
+import (
+	"encoding/binary"
+	"sync"
+
+	"github.com/jackc/pgtype"
+	"github.com/jackc/pgx/v4"
+	"github.com/timescale/promscale/pkg/log"
+)
+
+var fPool = sync.Pool{
+	New: func() interface{} {
+		return new(pgtype.Float8Array)
+	},
+}
+
+var tPool = sync.Pool{
+	New: func() interface{} {
+		return new(pgtype.TimestamptzArray)
+	},
+}
+
+//wrapper to allow DecodeBinary to reuse the existing array so that a pool is effective
+type timestamptzArrayWrapper struct {
+	*pgtype.TimestamptzArray
+}
+
+func (dstwrapper *timestamptzArrayWrapper) DecodeBinary(ci *pgtype.ConnInfo, src []byte) error {
+	dst := dstwrapper.TimestamptzArray
+	if src == nil {
+		*dst = pgtype.TimestamptzArray{Status: pgtype.Null}
+		return nil
+	}
+
+	var arrayHeader pgtype.ArrayHeader
+	rp, err := arrayHeader.DecodeBinary(ci, src)
+	if err != nil {
+		return err
+	}
+
+	if len(arrayHeader.Dimensions) == 0 {
+		*dst = pgtype.TimestamptzArray{Dimensions: arrayHeader.Dimensions, Status: pgtype.Present}
+		return nil
+	}
+
+	elementCount := arrayHeader.Dimensions[0].Length
+	for _, d := range arrayHeader.Dimensions[1:] {
+		elementCount *= d.Length
+	}
+
+	//reuse logic
+	elements := dst.Elements
+	if cap(dst.Elements) < int(elementCount) {
+		elements = make([]pgtype.Timestamptz, elementCount)
+	} else {
+		elements = elements[:elementCount]
+	}
+
+	for i := range elements {
+		elemLen := int(int32(binary.BigEndian.Uint32(src[rp:])))
+		rp += 4
+		var elemSrc []byte
+		if elemLen >= 0 {
+			elemSrc = src[rp : rp+elemLen]
+			rp += elemLen
+		}
+		err = elements[i].DecodeBinary(ci, elemSrc)
+		if err != nil {
+			return err
+		}
+	}
+
+	*dst = pgtype.TimestamptzArray{Elements: elements, Dimensions: arrayHeader.Dimensions, Status: pgtype.Present}
+	return nil
+}
+
+//wrapper to to allow DecodeBinary to reuse existing array so that a pool is effective
+type float8ArrayWrapper struct {
+	*pgtype.Float8Array
+}
+
+func (dstwrapper *float8ArrayWrapper) DecodeBinary(ci *pgtype.ConnInfo, src []byte) error {
```

Comment:
If the purpose of these funcs is to just reuse the `pgtype.Float8Array` allocations, why not just do this way:

```golang
func PutFloat8Array(ref *pgtype.Float8Array) {
    ref.ArrayDimensions = pgtype.ArrayDimensions{}
    ref.Set(nil)
    fPool.Put(ref)
}
```

Same applies for TimestamptzArray.

ref: https://github.com/jackc/pgtype/blob/2ca304d4617a72491b1844eb92e9ecd86f7b84e9/float8_array.go#L22